File: app/src/main/java/mb00/android/codehub/ui/repo/inject/RepoActivityModule.java
package mb00.android.codehub.ui.repo.inject;

import dagger.Module;
import dagger.Provides;
import mb00.android.codehub.ui.repo.interactor.RepoInteractor;
import mb00.android.codehub.ui.repo.interactor.RepoInteractorImpl;
import mb00.android.codehub.ui.repo.router.RepoRouter;
import mb00.android.codehub.ui.repo.router.RepoRouterImpl;
import mb00.android.codehub.ui.repo.view.RepoActivity;
import mb00.android.codehub.ui.repo.viewmodel.RepoViewModel;


@Module(subcomponents = {
        RepoCodeFragmentComponent.class,
        RepoCommitsFragmentComponent.class,
        RepoContributorsFragmentComponent.class,
        RepoIssuesFragmentComponent.class,
        RepoLicenseFragmentComponent.class,
        RepoPullRequestsFragmentComponent.class,
        RepoPulseFragmentComponent.class,
        RepoReadmeFragmentComponent.class,
        RepoReleasesFragmentComponent.class
})
public class RepoActivityModule {

    @Provides
    RepoInteractor provideRepoInteractor() {
        return new RepoInteractorImpl();
    }

    @Provides
    RepoRouter provideRepoRouter(RepoActivity activity) {
        return new RepoRouterImpl(activity);
    }

    @Provides
    RepoViewModel provideRepoViewModel(RepoInteractor interactor, RepoRouter router) {
        return new RepoViewModel(interactor, router);
    }

}

File: app/src/main/java/mb00/android/codehub/ui/repo/inject/RepoActivityModule.kt
package mb00.android.codehub.ui.repo.inject

import dagger.Module
import dagger.Provides
import mb00.android.codehub.api.manager.ApiCallManager
import mb00.android.codehub.ui.repo.interactor.RepoInteractor
import mb00.android.codehub.ui.repo.interactor.RepoInteractorImpl
import mb00.android.codehub.ui.repo.router.RepoRouter
import mb00.android.codehub.ui.repo.router.RepoRouterImpl
import mb00.android.codehub.ui.repo.view.RepoActivity
import mb00.android.codehub.ui.repo.viewmodel.RepoViewModel


@Module(includes = [RepoActivityFragmentsModule::class])
class RepoActivityModule {

    @Provides
    internal fun provideRepoInteractor(apiCallManager: ApiCallManager): RepoInteractor {
        return RepoInteractorImpl(apiCallManager)
    }

    @Provides
    internal fun provideRepoRouter(activity: RepoActivity): RepoRouter {
        return RepoRouterImpl(activity)
    }

    @Provides
    internal fun provideRepoViewModel(interactor: RepoInteractor, router: RepoRouter): RepoViewModel {
        return RepoViewModel(interactor, router)
    }

}
